main/src/ca/uwaterloo/flix/language/phase/EntryPoint.scala

```diff
@@ -0,0 +1,295 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package ca.uwaterloo.flix.language.phase
+
+import ca.uwaterloo.flix.api.Flix
+import ca.uwaterloo.flix.language.ast.{Ast, Scheme, SourceLocation, Symbol, Type, TypedAst}
+import ca.uwaterloo.flix.language.errors.EntryPointError
+import ca.uwaterloo.flix.language.phase.unification.ClassEnvironment
+import ca.uwaterloo.flix.util.Validation.{ToFailure, ToSuccess, flatMapN, mapN}
+import ca.uwaterloo.flix.util.{InternalCompilerException, Validation}
+
+/**
+  * Processes the entry point of the program.
+  *
+  * The entry point is replaced by a new function (`main%`) that calls the old entry point (`func`).
+  *
+  * If the argument to `func` has type `Array[String]`,
+  *   then the arguments are passed from `main%` to `func`.
+  * If the argument to `func` has type `Unit`,
+  *   then the arguments to `main%` are ignored.
+  * If the argument to `func` has some other type,
+  *   an error is raised.
+  *
+  * If the result type of `func` has type `Unit`,
+  *   then the result is returned from `main%` as normal.
+  * If the result type of `func` is some other type with a `ToString` instance,
+  *   then the result is printed in `main%`.
+  * If the result type of `func` is some other type without a `ToString` instance,
+  *   then an error is raised.
+  *
+  *  For example, given an entry point `func` with type `Array[String] -> Float64`,
+  *  we produce:
+  *  {{{
+  *  pub def main%(args: Array[String]): Unit = {
+  *      println(func(args))
+  *  }
+  *  }}}
+  */
+object EntryPoint {
+
+  /**
+    * The resulting scheme of the entry point function.
+    * `Array[String] -> Unit`
+    */
+  private val EntryPointScheme = Scheme(Nil, Nil, Type.mkImpureArrow(Type.mkArray(Type.Str, SourceLocation.Unknown), Type.Unit, SourceLocation.Unknown))
+
+  /**
+    * The default entry point in case none is specified. (`main`)
+    */
+  private val DefaultEntryPoint = Symbol.mkDefnSym("main")
+
+
+  def run(root: TypedAst.Root)(implicit flix: Flix): Validation[TypedAst.Root, EntryPointError] = flix.phase("EntryPoint") {
+    flatMapN(findOriginalEntryPoint(root)) {
+      // Case 1: We have an entry point. Wrap it.
+      case Some(entryPoint0) =>
+        mapN(visitEntryPoint(entryPoint0, root, root.classEnv)) {
+          entryPoint =>
+            root.copy(
+              defs = root.defs + (entryPoint.sym -> entryPoint),
+              entryPoint = Some(entryPoint.sym)
+            )
+        }
+      // Case 2: No entry point. Don't touch anything.
+      case None => root.toSuccess
+    }
+  }
+
+  /**
+    * Finds the entry point in the given `root`.
+    */
+  def findOriginalEntryPoint(root: TypedAst.Root)(implicit flix: Flix): Validation[Option[TypedAst.Def], EntryPointError] = {
+    root.entryPoint match {
```

Comment:
Maybe cleaner with just a double match and then

(Some, _) => ...
(None, Some()) => ...
_ => ....

or maybe not

Reply:
Since the second lookup depends on the first, we cannot do them in the at once.